### ISSUE TITLE
Refactoring KCP to support pluggable upgrade strategy

### ISSUE THREAD

**Issue by benmoss** [kubernetes-sigs/cluster-api#2919]
> ⚠️ Cluster API maintainers can ask to turn an issue-proposal into a [CAEP](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/YYYYMMDD-template.md) when necessary, this is to be expected for large changes that impact multiple components, breaking changes, or new large features.

***Goals***
1. Provide an abstraction for defining the current upgrade strategy of KCP such that we can later provide a way for users to select alternative strategies.
1. Clean up code in [KCP ControlPlane](https://github.com/kubernetes-sigs/cluster-api/blob/5fdcb9e8dcf0ce93d9eaf984a1de1b1377290f8c/controlplane/kubeadm/internal/control_plane.go#L97-L146)
1. Logic for when we scale up / down lives in one well unit-tested place

***Non-Goals/Future Work***
1. Implement alternative upgrade strategies

**User Story**

As a developer I would like to have a good abstraction around upgrade strategies so that I have confidence in the logic the system will perform.

**Detailed Description**

I've already begun spiking on this [here](https://github.com/benmoss/cluster-api/blob/edbca08213d6806a4a1979432d1a1fbd6d85e519/controlplane/kubeadm/internal/machine_collection.go#L259-L328). 

The idea is to create a `ScaleStrategy` interface that is responsible for deciding when and how to scale up and down the control plane.

```go
type ScaleStrategy interface {
	NeedsScaleUp() bool
	NeedsScaleDown() bool
	NextForScaleDown() *clusterv1.Machine
}
```

Our current strategy, best I can understand it, is:
- When the number of machines is less than the number of replicas, scale up is recommended
- When the number of machines is greater than the number of replicas, scale up is not recommended
- When the number of machines is equal to the number of replicas:
  - When machines are older than the KCP's UpgradeAfter field, scale up is recommended
  - When machines configuration hashes do not match the KCP's configuration hash, scale up is recommended
- When scale up is recommended, scale down is not recommended
- When the number of machines is greater than the number of replicas, scale down is recommended
- When scaling down, remove outdated machines before up-to-date machines
- When scaling down, remove machines from unknown failure domains before known failure domains
- When scaling down, remove machines from failure domains with the largest number of outdated machines first

One semi-radical aspect of this is it will mean we no longer differentiate upgrades from scale up / scale down at the controller level. This means that [all of the reconciliation](https://github.com/kubernetes-sigs/cluster-api/blob/5fdcb9e8dcf0ce93d9eaf984a1de1b1377290f8c/controlplane/kubeadm/controllers/upgrade.go#L55-L83) currently being done only when upgrading will need to be done on any scaling operation, but I don't think this should pose any problems as far as I understand it.

/kind proposal

related to #2688 and #2587 

@jan-est
@vincepri

**Comment by jan-est** [comment]
@benmoss idea of ​​the ```ScaleStrategy``` interface is good and makes it easy to implement alternative upgrade strategies. It is certainly a good idea to keep the ```DefaultScaleStrategy```` in such a way that the user cannot control it in any way and implement for example Rolling Update like features as their own strategies.

I'm not so sure about the scaling steps and could this cause a problem? @vincepri Do you have any thoughts that this could be a problem with scale in/out operations.

**Comment by benmoss** [comment]
We could definitely still alter this to have more methods to distinguish Upgrade scaling operations from regular reasons, it just means it'll be a broader interface. If it's feasible I think it's conceptually simpler to not special-case upgrading, but there might be more wrinkles there than I realize.

**Comment by vincepri** [comment]
I'm going to punt this to Next for now. This refactoring needs definitely more discussions. That said, we can start working on some implementation (or light refactoring) in the current codebase, but I would like to understand first what gains/improvements and outcomes we're trying to achieve.

/milestone Next

**Comment by detiber** [comment]
There are some potential issues that I see with allowing for additional upgrade strategies:
- How to handle a single control plane replica using the default of stacked etcd
- Impact on the failure tolerance of the etcd cluster during the upgrade process
- Ensuring the action we are taking would not cause the etcd cluster to lose quorum

By far the biggest driver for hardcoding the current strategy was due to the fact that we wanted to support the ability to have a KubeadmControlPlane consisting of only a single replica. With this requirement, the only way for us to upgrade the cluster (without resorting to backing up and restoring data) is to join an additional Machine to the cluster prior to removing the existing Machine.

The second major driver for the current strategy was to provide a level of safety during the upgrade operation around the failure tolerance capacity of the etcd cluster. Given the constraints we place on scaling (requiring an odd number of replicas when using the default stacked etcd) scaling up first affords us the ability to not reduce the failure capacity for the etcd cluster during upgrade operations.
- A 1 replica etcd cluster can not tolerate any failures, same as a 2 replica etcd cluster
- A 3 replica etcd cluster can tolerate a single failure, same as a 4 replica etcd cluster
The same cannot be said for scaling down prior to scaling up for upgrade, where we would reduce the failure tolerance during the upgrade process (https://etcd.io/docs/v3.4.0/faq/#what-is-failure-tolerance).

For completeness, we ruled out backing up and restoring data for a couple of reasons:
- The process would require special handling on the infrastructure provider side and we didn't want to impose additional requirements on the infrastructure provider.
- It is much more disruptive to the cluster during the upgrade process and any issues encountered would leave the cluster's control plane in a broken state for the duration of remediation.

In the past we had discussed the idea of adding additional checks around the current state of the etcd cluster and the potential impact it would have on the quorum of the cluster. If we were going to go down the path of configurable/pluggable upgrade strategies, I would think that adding additional safety protection around the protection of quorum would be a minimum requirement if we did decide to go down the path of alternative upgrade strategies, though.

**Comment by jan-est** [comment]
@detiber any extra safety in KCP would be good. Maybe we need to discuss about etcd backing up and restoring as well as additional checks on state of the etcd, before we can decide adding alternative upgrade strategies? Adding these feature into KCP is a major refactoring, though.

**Comment by vincepri** [comment]
As per April 27th 2020 community guidelines, this project follows the process outlined in https://github.com/kubernetes-sigs/cluster-api/blob/master/CONTRIBUTING.md#proposal-process-caep for large features or changes.

Following those guidelines, I'll go ahead and close this issue for now and defer to contributors interested in pushing the proposal forward to open a collaborative document proposal instead, and follow the process as described.

/close
